whelktool/scripts/cleanups/2025/05/lxl-4645-link-ispartof-redux.groovy

```diff
@@ -0,0 +1,251 @@
+/**
+ * Replace local isPartOf with link
+ * Some code borrowed from ../../2020/11/lxl-3379-link-isPartOf.groovy
+ * See https://kbse.atlassian.net/browse/LXL-4645
+ */
+
+String where = """
+    collection = 'bib' and deleted = false and data#>>'{@graph,1,isPartOf}' LIKE '%"controlNumber":%'
+"""
+
+skipped = getReportWriter("skipped")
+def whelk = getWhelk()
+
+selectBySqlWhere(where) { doc ->
+    def source_thing = doc.graph[1]
+    def _logSkip = { msg -> skipped.println("${doc.doc.getURI()}: ${msg}") }
+
+    List isPartOfs = asList(source_thing["isPartOf"])
+    if (isPartOfs.size() != 1) {
+        _logSkip("more than one isPartOf")
+        return
+    }
+    def isPartOf = isPartOfs[0]
+
+    def validSets = [
+        ["@type", "describedBy"] as Set,
+        ["@type", "describedBy", "identifiedBy"] as Set,
+        ["@type", "describedBy", "hasTitle"] as Set,
+        ["@type", "describedBy", "hasTitle", "identifiedBy"] as Set,
+    ]
+
+    if (!(validSets.any { it.equals(isPartOf.keySet()) })) {
+        _logSkip("too much stuff in isPartOf: ${isPartOf.keySet()}")
+        return
+    }
+
+    if (!(whelk.jsonld.isSubClassOf(isPartOf["@type"], "Instance"))) {
+        _logSkip("isPartOf.@type not Instance: ${isPartOf['@type']}")
+        return
+    }
+
+    if (isPartOf["describedBy"].size() != 1) {
+        _logSkip("more than one describedBy: ${isPartOf.describedBy}")
+        return
+    }
+
+    def describedBy = isPartOf["describedBy"][0]
+    if (!(describedBy instanceof Map && describedBy.keySet().equals(["@type", "controlNumber"].toSet()))) {
+        _logSkip("describedBy contains something other than [@type, controlNumber]: ${describedBy.keySet()}")
+        return
+    }
+    if (!(describedBy["controlNumber"] instanceof String)) {
```

Comment:
Could also be a single-element array (containing a string)? This goes for all places where a string value is expected.

Reply:
Looking at the logs there were 164 records (with local isPartOf) where controlNumber was a list, and in each case it had multiple elements, often with weird-looking stuff -- we can deal with stragglers like that later.
(And only one instance where hasTitle.mainTitle was not a string)